core/blockchain.go

```diff
@@ -2105,3 +2105,44 @@ func (bc *BlockChain) CXMerkleProof(shardID uint32, block *types.Block) (*types.
 	}
 	return proof, nil
 }
+
+// NextCXReceiptsProofUnspentCheckpoint returns the next checkpoint blockNum
+func (bc *BlockChain) NextCXReceiptsProofUnpentCheckpoint(currentNum uint64, shardID uint32) uint64 {
+	lastCheckpoint, _ := rawdb.ReadCXReceiptsProofUnspentCheckpoint(bc.db, shardID)
+	newCheckpoint := lastCheckpoint
+
+	// the new checkpoint will not exceed currentNum+1
+	for num := lastCheckpoint; num <= currentNum+1; num++ {
+		hash, _ := rawdb.ReadCXReceiptsProofUnspent(bc.db, shardID, num)
+		if hash == rawdb.EmptyHash {
+			// TODO: check if there is IncompingReceiptsHash in crosslink header
+			// if the rootHash is non-empty, it means incomingReceipts are not delivered
+			// otherwise, it means there is no cross-shard transactions for this block
+			newCheckpoint = num
```

Comment:
Seems these "newCheckpoint = num" can be grouped into one call.

Reply:
Since there is a TODO item, so I just make the separation of logic so that it is clearer to me. After finish TODO, we can optimize the code